From 28bad91f2dc8034f79b6ea896f3354817bc08982 Mon Sep 17 00:00:00 2001
From: Jeremy Kerr <jk@codeconstruct.com.au>
Date: Fri, 13 Oct 2023 17:28:22 +0800
Subject: [PATCH 2/2] mctp: make key lookups match the ANY address on either
 local or peer

We may have an ANY address in either the local or peer address of a
sk_key, and may want to match on an incoming daddr or saddr being ANY.

Do this by altering mctp_address_matches() to accept ANY in either
argument, and audit the non-key-match uses of it:

 - for lookup_prealloc_tag in the local output path, a peer destination
   address of ANY should match the preallocated tag too

 - for mctp_lookup_bind, an incoming broadcast should match the bind too

Reported-by: Eric Chuang <echuang@google.com>
Reported-by: Anthony <anthonyhkf@google.com>
Signed-off-by: Jeremy Kerr <jk@codeconstruct.com.au>
---
 include/net/mctp.h | 2 +-
 net/mctp/route.c   | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/net/mctp.h b/include/net/mctp.h
index 0c18a08c000d..e64acafc7817 100644
--- a/include/net/mctp.h
+++ b/include/net/mctp.h
@@ -55,7 +55,7 @@ static inline bool mctp_address_null(mctp_eid_t eid)
 
 static inline bool mctp_address_matches(mctp_eid_t match, mctp_eid_t eid)
 {
-	return match == eid || match == MCTP_ADDR_ANY;
+	return match == eid || match == MCTP_ADDR_ANY || eid == MCTP_ADDR_ANY;
 }
 
 static inline struct mctp_hdr *mctp_hdr(struct sk_buff *skb)
diff --git a/net/mctp/route.c b/net/mctp/route.c
index dac0fe149241..c675ccd00b24 100644
--- a/net/mctp/route.c
+++ b/net/mctp/route.c
@@ -79,7 +79,7 @@ static bool mctp_key_match(struct mctp_sk_key *key, mctp_eid_t local,
 	if (!mctp_address_matches(key->local_addr, local))
 		return false;
 
-	if (key->peer_addr != peer)
+	if (!mctp_address_matches(key->peer_addr, peer))
 		return false;
 
 	if (key->tag != tag)
-- 
2.39.2

